File: voice/intelligence/encryption/decrypt.java
import java.io.StringReader;
import java.nio.file.Files;
import java.nio.file.Path;
import java.security.KeyPair;
import java.security.PrivateKey;
import java.security.spec.MGF1ParameterSpec;
import java.util.Base64;
import javax.crypto.Cipher;
import javax.crypto.SecretKey;
import javax.crypto.spec.GCMParameterSpec;
import javax.crypto.spec.OAEPParameterSpec;
import javax.crypto.spec.PSource;
import javax.crypto.spec.SecretKeySpec;

import org.bouncycastle.openssl.PEMKeyPair;
import org.bouncycastle.openssl.PEMParser;
import org.bouncycastle.openssl.jcajce.JcaPEMKeyConverter;

public class DecryptScript {

    /*
     * CHANGE THESE VALUES TO YOURS
     */

    private static final String ENCRYPTED_KEY_B64 = "THIS_IS_THE_KEY_RETURNED_IN_THE_S3_HEADERS";
    private static final String IV_B64 = "THIS_IS_THE_IV_RETURNED_IN_THE_S3_HEADERS";
    private static final String CIPHERTEXT_FILE = "/path/to/encrypted/file";
    private static final String PRIVATE_KEY_FILE = "/path/to/private/key.pem";

    public static void main(final String[] args) throws Exception {
        // Load BC
        Security.addProvider(new BouncyCastleProvider());

        final byte[] encryptedKey = Base64.getDecoder().decode(ENCRYPTED_KEY_B64);
        final byte[] iv = Base64.getDecoder().decode(IV_B64);
        final byte[] cipherPlusTag = Files.readAllBytes(Path.of(CIPHERTEXT_FILE));

        final String privateKeyString = Files.readString(Path.of(PRIVATE_KEY_FILE));

        /* 2. RSA-decrypt (unwrap) the AES data key */
        final PrivateKey rsaPriv = buildEncryptionKeyPair(privateKeyString);

        final Cipher rsa = Cipher.getInstance("RSA/ECB/OAEPWithSHA-256AndMGF1Padding");
        final OAEPParameterSpec oaep = new OAEPParameterSpec(
            "SHA-256",
            "MGF1",
            MGF1ParameterSpec.SHA256,
            PSource.PSpecified.DEFAULT);
        rsa.init(Cipher.DECRYPT_MODE, rsaPriv, oaep);

        final byte[] rawAesKey = rsa.doFinal(encryptedKey);
        final SecretKey aesKey = new SecretKeySpec(rawAesKey, "AES");

        /* 3. AES/GCM decrypt the payload */
        final Cipher gcm = Cipher.getInstance("AES/GCM/NoPadding");
        final GCMParameterSpec gcmSpec = new GCMParameterSpec(128, iv);
        gcm.init(Cipher.DECRYPT_MODE, aesKey, gcmSpec);

        final byte[] plaintext = gcm.doFinal(cipherPlusTag);

        /* 4. Persist plaintext to disk */
        final String outFile = CIPHERTEXT_FILE + ".json";
        writeFile(outFile, plaintext);
        System.out.println("Decrypted data written to: " + outFile);
    }

    private static PrivateKey buildEncryptionKeyPair(final String privateKey) {
        try (PEMParser pemParser = new PEMParser(new StringReader(privateKey))) {
            JcaPEMKeyConverter converter = (new JcaPEMKeyConverter()).setProvider("BC");
            PrivateKeyInfo privateKeyInfo = (PrivateKeyInfo) pemParser.readObject();
            return converter.getPrivateKey(privateKeyInfo);
        } catch (Exception e) {
            throw new IllegalArgumentException("invalid private key", e);
        }
    }

    /**
     * Write bytes to a file, creating or replacing it.
     */
    private static void writeFile(final String path, final byte[] data) throws Exception {
        Files.write(Path.of(path), data);
    }
}

File: voice/intelligence/encryption/decrypt.py
#!/usr/bin/env python3
"""
Performs RSA-OAEP decryption of AES key, then AES-GCM decryption of payload.
"""

import base64
import os
from pathlib import Path
from typing import Union

from cryptography.hazmat.primitives import hashes, serialization
from cryptography.hazmat.primitives.asymmetric import padding
from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes
from cryptography.hazmat.backends import default_backend
from cryptography.hazmat.primitives.asymmetric.rsa import RSAPrivateKey


class DecryptScript:
    """
    CHANGE THESE VALUES TO YOURS
    """

    ENCRYPTED_KEY_B64 = "THIS_IS_THE_KEY_RETURNED_IN_THE_S3_HEADERS"
    IV_B64 = "THIS_IS_THE_IV_RETURNED_IN_THE_S3_HEADERS"
    CIPHERTEXT_FILE = "/path/to/encrypted/file"
    PRIVATE_KEY_FILE = "/path/to/your/private_key.pem"

    @staticmethod
    def main() -> None:
        """Main decryption process"""
        try:
            # Decode base64 values
            encrypted_key = base64.b64decode(DecryptScript.ENCRYPTED_KEY_B64)
            iv = base64.b64decode(DecryptScript.IV_B64)

            # Read encrypted file
            cipher_plus_tag = Path(DecryptScript.CIPHERTEXT_FILE).read_bytes()

            # Read private key
            private_key_string = Path(DecryptScript.PRIVATE_KEY_FILE).read_text()

            # 2. RSA-decrypt (unwrap) the AES data key
            rsa_private_key = DecryptScript._build_encryption_key_pair(private_key_string)

            # RSA/ECB/OAEPWithSHA-256AndMGF1Padding equivalent
            raw_aes_key = rsa_private_key.decrypt(
                encrypted_key,
                padding.OAEP(
                    mgf=padding.MGF1(algorithm=hashes.SHA256()),
                    algorithm=hashes.SHA256(),
                    label=None
                )
            )

            # 3. AES/GCM decrypt the payload
            # In AES-GCM, the authentication tag is typically appended to the ciphertext
            # We need to separate the ciphertext from the tag (last 16 bytes)
            ciphertext = cipher_plus_tag[:-16]
            tag = cipher_plus_tag[-16:]

            # Create AES-GCM cipher
            cipher = Cipher(
                algorithms.AES(raw_aes_key),
                modes.GCM(iv, tag),
                backend=default_backend()
            )
            decryptor = cipher.decryptor()

            # Decrypt the payload
            plaintext = decryptor.update(ciphertext) + decryptor.finalize()

            # 4. Persist plaintext to disk
            out_file = DecryptScript.CIPHERTEXT_FILE + ".json"
            DecryptScript._write_file(out_file, plaintext)
            print(f"Decrypted data written to: {out_file}")

        except Exception as e:
            print(f"Decryption failed: {e}")
            raise

    @staticmethod
    def _build_encryption_key_pair(private_key_string: str) -> RSAPrivateKey:
        """
        Parse private key from PEM string
        """
        try:
            private_key = serialization.load_pem_private_key(
                private_key_string.encode('utf-8'),
                password=None,
                backend=default_backend()
            )
            return private_key
        except Exception as ex:
            raise ValueError(f"Invalid private key: {ex}")

    @staticmethod
    def _write_file(path: Union[str, Path], data: bytes) -> None:
        """
        Write bytes to a file, creating or replacing it.
        """
        Path(path).write_bytes(data)


if __name__ == "__main__":
    DecryptScript.main()
